### PR TITLE
Feature/generics

### PR DESCRIPTION
**Pull request by j-m**
- Oaty can now be initialised with `new OatyArray<T>()` so that `.get` , `.data`, and `.transposed` return data of type `T`. This is completely optional as `T` defaults to `any`.
- Removed `missingKeyReturns` and `noResultsReturns` options.
- Exported a new type for the `_transposed` property.
- Expanded the ternary conditions to improve readability.
- Made the `data` constructor param optional and default to `[]`.
- When a key is not transposed and `.get` searches for that key, it will now throw a `ReferenceError`.
- Added NYC and increased tests to 100% coverage.

### CODE REVIEW

File: dist/index.d.ts [hunk 1]
```diff
@@ -1,18 +1,23 @@
-export interface Options {
+export interface IOptions {
```
Comment (jmsv) [inline dist/index.d.ts]: `IThing` is a C#y convention iirc - rename back to `Thing`

**Review by jmsv — APPROVED** [review]
lgtm

**Comment by jmsv** [comment]
@j-m think you missed an `IOptions`

**Comment by j-m** [comment]
Should I change `ci` to `npm run coverage` instead of `test` so Travis runs the coverage test too?

**Comment by jmsv** [comment]
@j-m if you want, yep

**Comment by jmsv** [comment]
lgtm! :shipit:

**Pull request merged**